Framework/Algorithms/src/EstimateScatteringVolumeCentreOfMass.cpp

```diff
@@ -0,0 +1,146 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2025 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MantidAlgorithms/EstimateScatteringVolumeCentreOfMass.h"
+#include "MantidAPI/InstrumentValidator.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidAPI/Run.h"
+#include "MantidAPI/Sample.h"
+#include "MantidGeometry/Instrument.h"
+#include "MantidGeometry/Objects/ShapeFactory.h"
+#include "MantidGeometry/Rasterize.h"
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/CompositeValidator.h"
+#include "MantidKernel/ListValidator.h"
+#include "MantidKernel/V3D.h"
+#include <unordered_map>
+
+namespace Mantid::Algorithms {
+
+using namespace API;
+using namespace Geometry;
+using namespace Kernel;
+
+namespace {
+const std::string UNIT_M = "m";
+const std::string UNIT_CM = "cm";
+const std::string UNIT_MM = "mm";
+static const std::unordered_map<std::string, double> unitToMeters{{UNIT_M, 1.0}, {UNIT_CM, 0.01}, {UNIT_MM, 0.001}};
+} // namespace
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(EstimateScatteringVolumeCentreOfMass)
+
+EstimateScatteringVolumeCentreOfMass::EstimateScatteringVolumeCentreOfMass()
+    : API::Algorithm(), m_inputWS(), m_cubeSide(0.0) {}
+
+void EstimateScatteringVolumeCentreOfMass::init() {
+
+  // The input workspace must have an instrument and units of wavelength
+  auto wsValidator = std::make_shared<CompositeValidator>();
+  wsValidator->add<InstrumentValidator>();
+
+  declareProperty(std::make_unique<WorkspaceProperty<>>("InputWorkspace", "", Direction::Input, wsValidator),
+                  "Input Workspace");
+  declareProperty(std::make_unique<PropertyWithValue<std::vector<double>>>("CentreOfMass", V3D(), Direction::Output),
+                  "Estimated centre of mass of illuminated sample volume");
+
+  auto moreThanZero = std::make_shared<BoundedValidator<double>>();
+  moreThanZero->setLower(1e-6);
+  declareProperty("ElementSize", 1.0, moreThanZero,
+                  "The size of one side of an integration element cube in {ElementUnits}");
+
+  std::vector<std::string> unitOptions{UNIT_M, UNIT_CM, UNIT_MM};
+  declareProperty("ElementUnits", UNIT_MM, std::make_shared<StringListValidator>(std::move(unitOptions)),
```

Comment:
Could just construct the vector as an R value reference if you want:
`std::make_shared<StringListValidator>(std::vector<std::string>{UNIT_M, UNIT_CM, UNIT_MM})`